File: src/main/java/wraith/fabricaeexnihilo/registry/crucible/CrucibleHeatRegistry.java
package wraith.fabricaeexnihilo.registry.crucible;

import com.google.common.reflect.TypeToken;
import net.minecraft.block.Block;
import net.minecraft.fluid.Fluid;
import net.minecraft.item.BlockItem;
import wraith.fabricaeexnihilo.api.recipes.crucible.CrucibleHeatRecipe;
import wraith.fabricaeexnihilo.compatibility.modules.MetaModule;
import wraith.fabricaeexnihilo.registry.AbstractRegistry;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class CrucibleHeatRegistry extends AbstractRegistry<List<CrucibleHeatRecipe>> implements wraith.fabricaeexnihilo.api.registry.CrucibleHeatRegistry {

    private final List<CrucibleHeatRecipe> registry;

    public CrucibleHeatRegistry() {
        this(new ArrayList<>());
    }

    public CrucibleHeatRegistry(List<CrucibleHeatRecipe> registry) {
        this.registry = registry;
    }

    @Override
    public void clear() {
        registry.clear();
    }

    @Override
    public boolean register(CrucibleHeatRecipe recipe) {
        return registry.add(recipe);
    }

    @Override
    public int getHeat(Block block) {
        return registry.stream().filter(recipe -> recipe.test(block)).findFirst().map(CrucibleHeatRecipe::value).orElse(0);
    }

    @Override
    public int getHeat(Fluid fluid) {
        return registry.stream().filter(recipe -> recipe.test(fluid)).findFirst().map(CrucibleHeatRecipe::value).orElse(0);
    }

    @Override
    public int getHeat(BlockItem item) {
        return registry.stream().filter(recipe -> recipe.test(item)).findFirst().map(CrucibleHeatRecipe::value).orElse(0);
    }

    @Override
    public void registerJson(File file) {
        if (file.exists()){
            try (var reader = new FileReader(file)) {
                List<CrucibleHeatRecipe> json = gson.fromJson(
                        reader,
                        SERIALIZATION_TYPE
                );
                json.forEach(this::register);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public List<CrucibleHeatRecipe> serializable() {
        return registry;
    }

    @Override
    public Collection<CrucibleHeatRecipe> getREIRecipes() {
        return registry;
    }

    private static final Type SERIALIZATION_TYPE = new TypeToken<List<CrucibleHeatRecipe>>(){}.getType();

    public static wraith.fabricaeexnihilo.api.registry.CrucibleHeatRegistry fromJson(File file) {
        return fromJson(file, wraith.fabricaeexnihilo.registry.crucible.CrucibleHeatRegistry::new, MetaModule.INSTANCE::registerCrucibleHeat);
    }

}

File: src/main/kotlin/exnihilofabrico/registry/crucible/CrucibleHeatRegistry.kt
package exnihilofabrico.registry.crucible

import com.google.gson.reflect.TypeToken
import exnihilofabrico.api.recipes.crucible.CrucibleHeatRecipe
import exnihilofabrico.api.registry.ICrucibleHeatRegistry
import exnihilofabrico.compatibility.modules.MetaModule
import exnihilofabrico.registry.AbstractRegistry
import net.minecraft.block.Block
import net.minecraft.fluid.Fluid
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import java.io.File
import java.io.FileReader
import java.lang.reflect.Type

data class CrucibleHeatRegistry(val registry: MutableList<CrucibleHeatRecipe> = ArrayList()): AbstractRegistry<MutableList<CrucibleHeatRecipe>>(), ICrucibleHeatRegistry {

    override fun clear() = registry.clear()

    override fun register(recipe: CrucibleHeatRecipe)  = registry.add(recipe)

    override fun getHeat(block: Block) = registry.firstOrNull { it.test(block) }?.value ?: 0
    override fun getHeat(fluid: Fluid) = registry.firstOrNull { it.test(fluid) }?.value ?: 0
    override fun getHeat(item: Item) = registry.firstOrNull { it.test(item) }?.value ?: 0
    override fun getHeat(stack: ItemStack) = registry.firstOrNull { it.test(stack) }?.value ?: 0

    override fun serializable() = registry
    override fun registerJson(file: File) {
        if(file.exists()){
            val json: MutableList<CrucibleHeatRecipe> = gson.fromJson(FileReader(file),
                SERIALIZATION_TYPE
            )
            json.forEach { register(it) }
        }
    }
    override fun getREIRecipes() = registry

    companion object {
        val SERIALIZATION_TYPE: Type = object: TypeToken<MutableList<CrucibleHeatRecipe>>(){}.type
        fun fromJson(file: File) = fromJson(
            file,
            { CrucibleHeatRegistry() },
            MetaModule::registerCrucibleHeat
        )
    }
}